File: src/model/message.ts
export class TGFSFileVersionSerialized {
  type: 'FV';
  id: string;
  updatedAt: number;
  messageId: number;
  size: number;
}

export class TGFSFileObject {
  type: 'F';
  name: string;
  versions: TGFSFileVersionSerialized[];
}

export class TGFSFileRefSerialized {
  type: 'FR';
  messageId: number;
  name: string;
}

export class TGFSDirectorySerialized {
  type: 'D';
  name: string;
  children: TGFSDirectorySerialized[];
  files: TGFSFileRefSerialized[];
}

File: tgfs/model/message.py
from typing import Literal, TypedDict


class TGFSFileVersionSerialized(TypedDict):
    type: Literal["FV"]
    id: str
    updatedAt: int
    messageId: int
    size: int


class TGFSFileObject(TypedDict):
    type: Literal["F"]
    name: str
    versions: list[TGFSFileVersionSerialized]


class TGFSFileRefSerialized(TypedDict):
    type: Literal["FR"]
    messageId: int
    name: str


class TGFSDirectorySerialized(TypedDict):
    type: Literal["D"]
    name: str
    children: list["TGFSDirectorySerialized"]
    files: list[TGFSFileRefSerialized]
